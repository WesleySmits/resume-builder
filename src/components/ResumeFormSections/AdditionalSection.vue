<template>
    <div class="form">
        <h1>Additional Information</h1>
        <p>Provide any additional information here.</p>
    </div>
</template>

<script setup lang="ts"></script>
