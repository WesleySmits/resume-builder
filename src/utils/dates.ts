/**
 * Formats a date into "MMM yyyy" (e.g., "Aug 2024").
 * @param date - The date to format.
 * @param locale - Optional locale string. Defaults to the user's browser locale.
 * @returns A string in the format "MMM yyyy".
 */
export function formatDateToLocale(passedDate: Date | string, locale?: string): string {
    const date = typeof passedDate === 'string' ? parseDateFromString(passedDate) : passedDate;
    const userLocale = locale || navigator.language || 'en-US';

    const options: Intl.DateTimeFormatOptions = {
        year: 'numeric',
        month: 'short',
    };

    return new Intl.DateTimeFormat(userLocale, options).format(date);
}

/**
 * Converts a string in "YYYY-MM-DD" format to a Date object.
 * @param dateString - The date string in "YYYY-MM-DD" format.
 * @returns A Date object or null if the input is invalid.
 */
export function parseDateFromString(dateString: string): Date {
    const [year, month, day] = dateString.split('-').map(Number);

    // Validate the parsed numbers
    if (!year || !month || !day || year < 1 || month < 1 || month > 12 || day < 1 || day > 31) {
        throw new Error(`Invalid date string: ${dateString}`);
    }

    // Create a Date object (note: month is zero-based in JavaScript Date)
    return new Date(year, month - 1, day);
}
