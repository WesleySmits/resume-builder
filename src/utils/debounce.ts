/* eslint-disable @typescript-eslint/no-explicit-any */
// eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
export function debounce(func: Function, wait: number = 750) {
    let timeout: number | undefined
    return function (this: any, ...args: any[]) {
        clearTimeout(timeout)
        timeout = window.setTimeout(() => func.apply(this, args), wait)
    }
}
