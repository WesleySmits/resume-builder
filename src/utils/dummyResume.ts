export function getDummyResume(): ResumeData {
    return {
        general: {
            name: {
                firstName: 'Jon',
                middleName: '',
                lastName: 'Snow',
                displayName: '',
            },
            profilePhoto: 'data:image/jpeg;base64,',
            region: 'The Wall',
            contact: {
                email: 'jon.snow@resume-maker.io',
                phone: '123123123',
            },
            drivingLicense: 'Car',
            functionTitle: 'Watch Commander',
            achievements: ['Defeated the Night King.. \n and his dragon.. \n Sort of', '', 'King in the North'],
            introduction:
                'I am the sword in the darkness. \n I am the watcher on the walls. \n I am the shield that guards the realms',
            colleaguesDescribe: 'Brave',
            colleaguesKnow: 'Loyal',
        },
        skills: {
            languages: ['HTML', 'CSS', 'JavaScript'],
            frameworks: ['Vue.js'],
            platforms: ['Node.js'],
            methodologies: ['Agile'],
            databases: ['MongoDB'],
            tools: ['Git'],
            operatingSystems: ['MacOS'],
        },
        topSkills: [
            {
                name: 'JavaScript',
                yearsOfExperience: 10,
            },
            {
                name: 'TypeScript',
                yearsOfExperience: 6,
            },
        ],
        education: [],
        certifications: [],
        jobs: [],
        personalProjects: [
            {
                title: 'Project 1',
                description: 'This was a very tough 5-minute project.',
                period: {
                    startDate: '2001-01-01',
                    endDate: '2023-01-01',
                },
                skills: {
                    languages: ['JavaScript', 'TypeScript'],
                    frameworks: ['Vue.js', 'React'],
                    platforms: ['Node.js', 'Docker'],
                    methodologies: ['Agile', 'Scrum'],
                    databases: ['MongoDB', 'PostgreSQL'],
                    tools: ['Git', 'Docker'],
                    operatingSystems: ['Windows', 'Linux'],
                },
            },
            {
                title: 'Project 2',
                description: 'This was a very tough 5-minute project.',
                skills: {
                    languages: ['JavaScript', 'TypeScript'],
                    frameworks: ['Vue.js', 'React'],
                    platforms: ['Node.js', 'Docker'],
                    methodologies: ['Agile', 'Scrum'],
                    databases: ['MongoDB', 'PostgreSQL'],
                    tools: ['Git', 'Docker'],
                    operatingSystems: ['Windows', 'Linux'],
                },
            },
        ],
    };
}

export function setupDummyResume(): ResumeData {
    const resumeInitialState = getDummyResume();
    localStorage.setItem('resumeData', JSON.stringify(resumeInitialState));
    return resumeInitialState;
}
