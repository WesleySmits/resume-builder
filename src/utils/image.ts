export function base64ToUint8Array(base64: string): Uint8Array {
    const binaryString = atob(base64.split(',')[1]);
    const len = binaryString.length;
    const bytes = new Uint8Array(len);
    for (let i = 0; i < len; i++) {
        bytes[i] = binaryString.charCodeAt(i);
    }
    return bytes;
}

export function roundImage(base64: string, width: number, height: number): Promise<string> {
    return new Promise((resolve) => {
        const img = new Image();
        img.src = base64;
        img.onload = () => {
            const canvas = document.createElement('canvas');
            canvas.width = width;
            canvas.height = height;
            const ctx = canvas.getContext('2d');

            if (ctx) {
                ctx.clearRect(0, 0, width, height);
                ctx.beginPath();
                ctx.arc(width / 2, height / 2, Math.min(width, height) / 2, 0, Math.PI * 2, true);
                ctx.closePath();
                ctx.clip();

                ctx.drawImage(img, 0, 0, width, height);

                resolve(canvas.toDataURL());
            }
        };
    });
}

export async function convertImageToBase64(file: File): Promise<string> {
    return new Promise((resolve, reject) => {
        try {
            const reader = new FileReader();
            reader.onload = () => resolve(reader.result as string);
            reader.onerror = () => reject(new Error('Failed to convert file to Base64'));
            reader.readAsDataURL(file); // This is where we can inject an error
        } catch (error) {
            reject(new Error(error as string));
        }
    });
}
