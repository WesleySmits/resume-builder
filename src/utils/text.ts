export function capitalize(text: string): string {
    if (!text) {
        return '';
    }

    return text.charAt(0).toUpperCase() + text.slice(1).toLowerCase();
}

export function toCamelCase(text: string): string {
    return text
        .replace(/[^a-zA-Z0-9]/g, ' ')
        .split(' ')
        .map((word, index) => (index === 0 ? word : capitalize(word)))
        .join('');
}
