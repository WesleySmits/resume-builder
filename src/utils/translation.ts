import * as en from '@/locales/en.json';
import * as nl from '@/locales/nl.json';

const locales: Record<Locales, Record<string, string>> = {
    en: en,
    nl: nl,
};

export function getLocalizedString(
    key: string,
    variables?: Record<string, string | number>,
    lang: Locales = 'en',
): string {
    const translationString = locales[lang][key];

    if (variables) {
        return (
            translationString?.replace(/{(.*?)}/g, (_, match) => {
                return variables[match] as string;
            }) || key
        );
    }

    return translationString || key;
}
