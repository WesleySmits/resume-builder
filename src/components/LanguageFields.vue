<template>
    <DynamicList
        :title="getLocalizedString('languages')"
        :items="languages"
        :default-item="defaultLanguage"
        :onUpdate="updateLanguages"
        :getKey="getKey"
        :direction="'column'"
        :id="'languages'"
    >
        <template #item-fields="{ item, index, updateField, removeItem }">
            <fieldset :id="`language-fieldset-${index}`">
                <details name="language-curtain" :open="index === 0">
                    <summary>
                        <header>
                            <legend>{{ item.name || getLocalizedString('newLanguage') }}</legend>
                        </header>
                    </summary>

                    <div class="form">
                        <FormField
                            :type="'text'"
                            :required="true"
                            :id="`language-${index}`"
                            :label="getLocalizedString('language')"
                            :placeholder="getLocalizedString('languagePlaceholder')"
                            :helperText="getLocalizedString('languageHelperText')"
                            :modelValue="item.name"
                            @valid="(value) => updateField('name', value as string)"
                        />

                        <FormField
                            :type="'select'"
                            :required="true"
                            :id="`language-experience-${index}`"
                            :label="getLocalizedString('languageExperience')"
                            :helperText="getLocalizedString('languageExperienceHelperText')"
                            :placeholder="getLocalizedString('languageExperiencePlaceholder')"
                            :modelValue="item.experience"
                            @valid="(value) => updateField('experience', value as string)"
                            :options="[
                                { value: 'Beginner', text: getLocalizedString('beginner') },
                                { value: 'Intermediate', text: getLocalizedString('intermediate') },
                                { value: 'Advanced', text: getLocalizedString('advanced') },
                                { value: 'Fluent/Native', text: getLocalizedString('fluent') },
                            ]"
                        />

                        <button @click="removeItem" type="button" class="secondary" data-action="remove">
                            {{ getLocalizedString('removeLanguage') }}
                        </button>
                    </div>
                </details>
            </fieldset>
        </template>
    </DynamicList>
</template>

<script setup lang="ts">
import { getLocalizedString } from '@/utils/translation';
import DynamicList from './DynamicList.vue';
import { useResumeStore } from '@/stores/resume';
import FormField from './FormField.vue';

const store = useResumeStore();
const languages = store.languages;

function defaultLanguage(): Language {
    return {
        name: '',
        experience: 'Beginner',
    };
}

function getKey(item: Language, index: number): string {
    return `language-${item.name}-${index}`;
}

function updateLanguages(newLanguages: Language[]): void {
    store.setLanguages(newLanguages);
}
</script>
