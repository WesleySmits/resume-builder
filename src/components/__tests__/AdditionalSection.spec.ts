import { mount, VueWrapper } from '@vue/test-utils';
import { describe, it, expect, vi } from 'vitest';
import AdditionalSection from '../ResumeFormSections/AdditionalSection.vue';
import { createTestingPinia } from '@pinia/testing';
import { getDummyResume } from '@/utils/dummyResume';
import LanguageFields from '../LanguageFields.vue';

describe('AdditionalSection.vue', () => {
    const resumeInitialState = getDummyResume();

    function getMountedComponent(): VueWrapper {
        return mount(AdditionalSection, {
            global: {
                plugins: [
                    createTestingPinia({
                        createSpy: vi.fn,
                        initialState: {
                            resume: resumeInitialState,
                        },
                    }),
                ],
            },
        });
    }

    it('renders correctly', () => {
        const wrapper = getMountedComponent();
        expect(wrapper.exists()).toBe(true);
    });

    it('renders the correct components', () => {
        const wrapper = getMountedComponent();
        expect(wrapper.findComponent(LanguageFields).exists()).toBe(true);
    });

    it('matches the snapshot', () => {
        const wrapper = mount(AdditionalSection);
        expect(wrapper.html()).toMatchSnapshot();
    });
});
