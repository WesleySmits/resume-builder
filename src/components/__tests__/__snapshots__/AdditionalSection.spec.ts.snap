// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AdditionalSection.vue > matches the snapshot 1`] = `
"<div class="form">
  <fieldset data-v-7fdf210c="" class="dynamic-list" id="languages">
    <details data-v-7fdf210c="" name="skills-curtain" open="">
      <summary data-v-7fdf210c="">
        <header data-v-7fdf210c="">
          <legend data-v-7fdf210c="">Languages</legend>
          <div data-v-7fdf210c="" class="buttons"><button data-v-7fdf210c="" class="primary icon" data-action="add"><svg data-v-7fdf210c="" xmlns="http://www.w3.org/2000/svg" fill="none" stroke="currentColor" viewBox="0 0 24 24">
                <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M12 6v6m0 0v6m0-6h6m-6 0H6"></path>
              </svg></button>
            <!--v-if-->
          </div>
        </header>
      </summary>
      <ul data-v-7fdf210c="" class="dynamic-list__items">
        <li data-v-7fdf210c="" class="list-item list-item--column">
          <fieldset id="language-fieldset-0">
            <details name="language-curtain" open="">
              <summary>
                <header>
                  <legend>English</legend>
                </header>
              </summary>
              <div class="form">
                <div data-v-b9645e84="" class="form-field"><input data-v-b9645e84="" fieldtype="" id="language-0" id-prop="" class="form-field__input" placeholder="English, Dutch, Spanish, etc." required="" aria-labelledby="language-0" aria-describedby="language-0-helper" modelvalue="English" type="text" value="English"><label data-v-b9645e84="" for="language-0">Language</label><small data-v-b9645e84="" id="\`\${id}-helper\`" class="help-text">Enter the language you speak.</small>
                  <!--v-if-->
                </div>
                <div data-v-b9645e84="" class="form-field"><select data-v-b9645e84="" fieldtype="" id="language-experience-0" id-prop="" class="form-field__input" placeholder="languageExperiencePlaceholder" required="" aria-labelledby="language-experience-0" aria-describedby="language-experience-0-helper" modelvalue="Fluent/Native" value="Fluent/Native">
                    <option data-v-b9645e84="" data-test="Fluent/Native" value="">languageExperiencePlaceholder</option>
                    <option data-v-b9645e84="" value="Beginner">Beginner</option>
                    <option data-v-b9645e84="" value="Intermediate">Intermediate</option>
                    <option data-v-b9645e84="" value="Advanced">Advanced</option>
                    <option data-v-b9645e84="" selected="" value="Fluent/Native">Fluent / Native</option>
                  </select><label data-v-b9645e84="" for="language-experience-0">languageExperience</label><small data-v-b9645e84="" id="\`\${id}-helper\`" class="help-text">languageExperienceHelperText</small>
                  <!--v-if-->
                </div><button type="button" class="secondary" data-action="remove">Remove Language</button>
              </div>
            </details>
          </fieldset>
        </li>
      </ul>
    </details>
  </fieldset>
</div>"
`;
